### DISCUSSION TITLE
Using Manifold-3d npm package from node?

### DISCUSSION THREAD

**Discussion by fearthecowboy** [elalish/manifold#754] [Ideas]
Is it possible to use the [manifold-3d](https://www.npmjs.com/package/manifold-3d) package as part of a nodejs (or [Deno](https://deno.com/))  app?

I played around with the package a bit, and didn't really get very far. 

I couldn't really even tell if it is possible to run it in node - the typescript typings are fighting me a bit, and before I dove in too deep, I thought I'd see if I'm even in the ballpark.

I really want to be able to make 3d models in TypeScript and then use a CLI to spit out 3MF (or ... STL or something else that the slicer will like) 

**Background:** 
My ultimate goal is to be able to construct 3d models for printing.  And to have more programmatic control over them than OpenSCAD alone gives me. 

I've done a considerable amount of work with OpenSCAD in the past (and, by the way, manifest made *that* a metric brazillion better!) - but the OpenSCAD syntax has never been my favorite. I poked around looking for a TypeScript/JavaScript front end that would do what I wanted, and found a couple efforts, but wasn't *really* satisfied with the direction or style.  Yeah, I'm picky 😄  (and I really don't much enjoy python, in spite of its many benefits)

So a couple weeks back I started prototyping my own "Tscad" for the hell of it, and made some really good progress, and then I wandered across https://manifoldcad.org/ ... which was surprisingly nice. 

I'm not married to outputting OpenSCAD - although, it's pretty solid for so many things. I'm pretty sure that I'd be just as happy using Manifold-3d directly if I could get it to do what I'd like.  

And then I think about all the nice libraries from OpenSCAD and either I'd have to implement a transpiler to go from that to TS. Hmm.  

thoughts?

**Comment by elalish** [comment]
You and I are well-aligned! First, did you see that ManifoldCAD can export 3MF already in addition to GLB? I wanted that same TS -> 3MF path as you. 

Also, yes, I can't see why our npm package wouldn't work in nodejs, though I haven't personally tested it. Would love to know what troubles you've had.

The main thing I feel ManifoldCAD is lacking is the ability to import libraries (e.g. involute gears, screws, etc) and existing manifold meshes. Of course Manifold is perfectly capable, but the web doesn't have great local file access and I don't feel like setting up a server. I have an [idea](https://github.com/elalish/manifold/discussions/727#discussioncomment-8294482) to package ManifoldCAD into a VSCode extension to make it a lot more powerful (and more discoverable). I'd love to find collaborators for this project and it sounds like you have the right skill set. Let me know if you're interested!

**Reply by fearthecowboy** [reply]
Yes, I'd love to work on this. (Luckily, it's right in the middle of my skillset -- my day job is currently working on the VSCode C++ extension, and previously several other ones) -- The only thing I'm not well-versed in is 3d math 😆 ... but I got the feeling that's not something I really need here...

I'd love to make sure that there are CLI tools so that it can be automated as well. Wiring up an extension in VSCode after that is pretty easy. 

(The irony, is that in addition to my day job, this week I already started another project doing a VSCode extension. Lucky for me I love TypeScript.)

>  did you see that ManifoldCAD can export 3MF already 

Yes, I saw that -- which is why I first started looking at where it's going. 

> Oh, actually, our CI runs our npm package's tests in Node, so it definitely works!

DOH! I never even thought to look at the CI jobs -- that will help me get started tremendously. I can rewrite/update the `.d.ts` bindings then to make it work with regular node/deno style development.

I do think it should be possible to build some kind of transpiler for OpenSCAD source files to spit something out that works with this too. 

Lemme spend a few more days looking at this. Now that I know that the code does run in node, this shouldn't be too much trouble.

**Comment by elalish** [comment]
Oh, actually, our CI runs our npm package's tests in Node, so it definitely works!

**Comment by pca006132** [comment]
maybe we should convert this to a discussion?

**Comment by fearthecowboy** [comment]
Wait ... you did 3dBuilder? Wow. 

A week ago I went on a hunt for the source for that and found the repo internally at work, I was seriously considering seeing if I could hack that to add an OpenSCAD-like (but TypeScript) scripting interface to that. 

I guess no point in going down that rabbit hole 😄

**Reply by fearthecowboy** [reply]
Well, I did get the code to run under node/deno - I was so close before, just didn't realize how close I was. 

I'm working on isolated the 3mf-saving code from to3mf.ts file so that it's not tied to the browser/sample.

**Reply by elalish** [reply]
Lol! I can't take credit for any of the code - it was all written by an excellent Slovakian vendor team, which had been a CAD startup. That was back when I was a PM, before I knew any C++, so I was managing their contract and specing features for them to build. 

You might also find Netfabb's mesh repair code in there - we paid them to rewrite it for us in C++ before Autodesk acquired them. That's still the best general manifold repair solution I've seen. 

What do you do at M$?

**Reply by fearthecowboy** [reply]
Currently, I'm in the MSVC org working on C++ design time features for VSCode and VS. I'm tasked with improving the performance of the IntelliSense and browse operations in VSCode. I *was* on the team working on building a better development experience for embedded developers, but that effort got "paused" and I got re-org'd onto the C++ design time team.  My job is to pretty much keep digging until I find something to fix and fix it. ([recently I fixed the goto symbol (ctrl-c) for C++](https://github.com/microsoft/vscode-cpptools/issues/11557) -- it's a lot faster now).

That Netfabb mesh repair is absolute magic. I was thinking that I wanted to extract that code out into a command line tool, just for fun.  I laugh every time I see some 3d printing youtuber bring it up as "the tool" for doing that, and the reason to keep 3dbuilder around. 

I too started off as a PM 18+ years ago... and then discovered that I'll only be happy with code dripping thru my fingers.  I tell people that "I'll die with my keyboard in my hands waiting to have my last PR merged"

**Reply by elalish** [reply]
Well, you work on the best product M$ ever made, so not bad. Once upon a time Netfabb's repair became some kind of Windows cloud service, but I'm sure that got turned down ages ago. Go look up Kris Iverson sometime - I think he's still kicking around there somewhere. He's a true believer when it comes to 3D printing.

Agreed, the power to build is true power indeed. If you're here in the greater Seattle area somewhere we should totally get lunch sometime!

**Comment by yacineMTB** [comment]
Hi! I'm poking around at this too. What I want and will probably end up with: a model viewer that updates on save from definitions on node. When I have something ergonomic I'll share here

**Reply by yacineMTB** [reply]
https://gist.github.com/yacineMTB/d28a67da85ed65583726cbc9ec9fe61f

1. build the wasm project such that you can run bindings/wasm/examples
2. paste the gist into `wew.ts`
3. install tsx
4. run the script with `npx tsx wew.ts`
5. enjoy your glbs

If i have more time tonight i'll set up a workshop repo with a GLB renderer. I might write my own

**Reply by elalish** [reply]
Nice! If you're looking for a GLB renderer, consider my other project: [`<model-viewer>`](https://modelviewer.dev/)

**Reply by elalish** [reply]
Also, I've been thinking it might be nice to publish worker.ts and some of the related files into an npm package. Would that be helpful?

**Reply by yacineMTB** [reply]
Thank you for your hard work, it's incredibly useful to me. Yes, I'll likely just wrap around <model-viewer> and figure out how to hack live reloading 

Re worker.ts, honestly, consider just shipping it along manifold-3D? What would help other folks without my elbow grease are quick examples they can just run and get a glb out of. If you're too busy I can help!

**Reply by elalish** [reply]
Yes, help would be much appreciated! I'm still surprised by how little I know of how people consume npm packages. If you have ideas on how best to make our code available for use, I'm open to them.

**Reply by yacineMTB** [reply]
ack, let me cook for a week or so and put up a diff

**Reply by a19grey** [reply]
blast from the past, I'm using DingCAD and stumbled on this trying to figure out how to import and STL file and render in DingCAD.  Just getting started and currently stuck at line one of the code... [1] SyntaxError: The requested module '../manifold_lib/built/manifold' does not provide an export named 'Mesh'.

**Reply by elalish** [reply]
What version are you using and what language? For v3.0 we removed `Mesh` in favor of `MeshGL`. Also note that STLs are often not very manifold-friendly and tend to require `Merge()`, which is heuristic.

**Reply by a19grey** [reply]
Thanks! I'm using [DingCAD](https://github.com/yacineMTB/dingcad/tree/master/manifold_lib) which is @yacineMTB project from this thread. I.. couldn't figure out what version of manifold cad it uses, looks like it's a specific custom build version ::shrug::.   I'm wildly out of my technical depth here.

I often have to import as .stl since many 3D printing project files only share this version (3mf is becoming more popular). OpenSCAD imports it well, but I really wanted to do CAD inside my existing IDE where I can leverage LLM's etc.

**Reply by a19grey** [reply]
OK re-reading DingCAD's readme "manifold_lib was ripped from manifoldcad and I was too lazy to detypescript it" so I think it's not a full ManifoldCad release since he ran into issues making it handle the live-reload features. 

Which is surprising since https://manifoldcad.org exists and I'm too ignorant to understand why it's challenging to make what is basically a local version that is the equivalent of clicking the 'run' button each time I save a file.

**Reply by elalish** [reply]
I've been wanting to repackage ManifoldCAD as a VSCode extension - would that be of use?

**Reply by pca006132** [reply]
Probably, I don't understand why hot reload is a problem as well. I think we can easily change our editor script to make it consume a local file.

**Reply by AlexKTracerMain** [reply]
VS Code extension would work great for many use-cases including mine!

And i want to be clear i am very ignorant here... and started on this journey pretty recently...

BUT my sense is that an easier to install package (like npm) could achieve a lot of the same functionality but with wider application so then embedding ManifoldCad in a website/app/etc would be easier.... if thats what you want.

Whereas it seems it would take a fair bit of VS-specific coding to make an extension 🤷‍♂️

**Reply by pca006132** [reply]
I think live reload doesn't really require vscode specific code, maybe using node API to run a backend for manifoldcad, reload the script and run it automatically when the file is changed will be good enough.

**Reply by hrgdavor** [reply]
Hi, I am personally committed to improve code-cad in JavaScript sphere. I prefer to code my 3d models for printing in JavaScript. Other people might prefer a different language, and that would lead them to other code-cad options and/or other manifold bindings.



https://github.com/hrgdavor/jscadui/ is a place where I am trying to provide building blocks for others, and there is a prototype app https://jscad.app . Sadly, it is a slow walk, but it is still going. 

It has a worker that can run local code and refresh preview when you save. for single script drag and drop it into the window. For multifile projects, drag&drop the base folder. 
 
The worker will happily load old js based on require, esm with imports and also typescript and you can even mix them in the same project. The prototype is made to run @jscad/modeling scripts, but manifold integration is ongoing. there is separate branch for it. 
 
There is a discord channel for those that want to chat on the topic on code-cad, manifold and some others https://discord.gg/6PB7qZ4HC7

**Reply by elalish** [reply]
@AlexKTracerMain Agreed, and publishing our manifoldCAD worker to npm would be a pre-requisite for making a VSCode extension anyway. I think it's actually pretty easy to publish it to npm - I just haven't sat down to do it. Maybe it's time.
